### PR TITLE
Implement target-typed conditional expressions.

### PR DESCRIPTION
**Pull request by gafter**
Also incidentally removes some cascaded diagnostics in definite assignment.
Also adds the default literal to the set of forms that require conversion to a type.

For spec, see https://github.com/dotnet/csharplang/pull/3363

### CODE REVIEW

File: src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml [hunk 1]
```diff
@@ -505,6 +513,9 @@
     <Field Name="Alternative" Type="BoundExpression"/>
 
     <Field Name="ConstantValueOpt" Type="ConstantValue?"/>
+
+    <Field Name="NaturalTypeOpt" Type="TypeSymbol?" Null="allow"/>
+    <Field Name="WasTargetTyped" Type="bool" />
```
Comment (alrz) [inline src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: I'm going to use this changes as a baseline for new() so we're consistent over all target-typing features.

Comment (gafter) [reply src/Compilers/CSharp/Portable/BoundTree/BoundNodes.xml]: I'll probably rebase this PR.  It is still under development.

**Comment by gafter** [comment]
For those of you on this channel, I would like to consider adding the following forms to those that require a conversion to a type:
- Method group
- Lambda expression
- `null` literal (except in specific contexts)
- throw expression
- tuple expression (when it has no natural type)

That work doesn't belong on this branch, but is something we could do later after merging to master. 
It could likely simplify the compiler a lot because we have to handle these kinds of bound nodes specially in lots of different places in the code.

**Comment by jcouv** [comment]
@gafter Sorry, some paperwork:
- Is the description (paragraph spec) in https://github.com/dotnet/csharplang/issues/2460 representative for this PR?
- I couldn't find a test plan issue for the feature. Could you create one and also add the feature to feature status page with links?

Thanks

**Comment by gafter** [comment]
@jcouv 
> Is the description (paragraph spec) in dotnet/csharplang#2460 representative for this PR?

Yes, with the exception of an issue to be discussed in tomorrow's LDM. That is, in this PR a conditional expression that has a common type is not a candidate to be target-typed.  That change from the spec in dotnet/csharplang#2460 is necessary to not break compatibility, and if accepted by the LDM will be updated in the description there.

> I couldn't find a test plan issue for the feature. Could you create one and also add the feature to feature status page with links?

Sure, in a separate PR. As soon as @jaredpar assigns reviewers for this feature (which he said he'll do imminently) I'll update the feature status page and ask them to start a test plan.

**Comment by gafter** [comment]
@dotnet/roslyn-compiler May I please have a couple of reviews of this implementation of target typing for conditional expressions?

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 2]
```diff
@@ -2190,18 +2190,32 @@ internal BoundExpression GenerateConversionForAssignment(TypeSymbol targetType,
...
+
+                        Debug.Assert(reportedError);
+                        return;
+                    }
+                case BoundKind.UnconvertedConditionalOperator:
+                    {
+                        var conditionalOperator = (BoundUnconvertedConditionalOperator)operand;
+                        HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+                        bool reportedError = false;
+                        foreach (var side in new[] { conditionalOperator.Consequence, conditionalOperator.Alternative })
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:2203]: Consider just writing these out, there's no need to allocate an array just for this.

File: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs [hunk 3]
```diff
@@ -2190,18 +2190,32 @@ internal BoundExpression GenerateConversionForAssignment(TypeSymbol targetType,
...
                         foreach (var arm in switchExpression.SwitchArms)
-                        {
-                            var armConversion = this.Conversions.ClassifyImplicitConversionFromExpression(arm.Value, targetType, ref useSiteDiagnostics);
-                            if (!armConversion.IsImplicit || !armConversion.IsValid)
-                            {
-                                GenerateImplicitConversionError(diagnostics, arm.Value.Syntax, armConversion, arm.Value, targetType);
-                                reportedError = true;
-                            }
-                        }
+                            tryConversion(arm.Value, ref reportedError, ref useSiteDiagnostics);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:2193]: Consider wrapping in braces. All other one-liners in this function do, including a `while` loop above.

**Comment by 333fred** [comment]
>             // invocation of operator true.

Entirely unrelated to this PR: it may be bound that way, but it's not actually run that way. If dynamic object implements both an implicit conversion to bool and operator true, the implicit conversion is used at runtime.

---
Refers to: src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs:2316 in 2a2135d. [](commit_id = 2a2135dcf70f4a1f029113b27761dd64303bbab5, deletion_comment = False)

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs [hunk 4]
```diff
@@ -0,0 +1,246 @@
...
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    /// <summary>
+    /// Test binding of the target-typed conditional (aka ternary) operator.
+    /// </summary>
+    public class TargetTypedConditionalOperatorTests : CSharpTestBase
+    {
+        [Fact]
+        public void TestImplicitConversions_Good()
```
Comment (RikkiGibson) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs:25]: nit: it seems like this test contains some "bad" cases, such as the first case that implicitly converts int to short. Was that intended?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 5]
```diff
@@ -258,6 +258,9 @@ internal BoundExpression BindToTypeForErrorRecovery(BoundExpression expression,
         /// </summary>
         internal BoundExpression BindToNaturalType(BoundExpression expression, DiagnosticBag diagnostics, bool reportDefaultMissingType = true)
         {
+            if (!expression.NeedsToBeConverted())
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:261]: Weren't we asserting that all expressions ended up converted and that `WithWasConverted` was called?

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 6]
```diff
@@ -275,6 +278,38 @@ internal BoundExpression BindToNaturalType(BoundExpression expression, Diagnosti
...
+                            if (op.NoCommonTypeError == ErrorCode.ERR_InvalidQM)
+                            {
+                                if (trueArg is Symbol trueSymbol && falseArg is Symbol falseSymbol)
+                                {
+                                    SymbolDistinguisher distinguisher = new SymbolDistinguisher(this.Compilation, trueSymbol, falseSymbol);
+                                    trueArg = distinguisher.First;
+                                    falseArg = distinguisher.Second;
+                                }
+
+                                diagnostics.Add(ErrorCode.ERR_InvalidQM, op.Syntax.Location, trueArg, falseArg);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:301]: >ERR_InvalidQM [](start = 58, length = 13)

I know you didn't introduce these error codes in this PR, but this is a really bad name. Consider either commenting to say what it is or changing the constant name.

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs [hunk 7]
```diff
@@ -275,6 +278,38 @@ internal BoundExpression BindToNaturalType(BoundExpression expression, Diagnosti
...
+                            }
+                            else
+                            {
+                                Debug.Assert(op.NoCommonTypeError == ErrorCode.ERR_AmbigQM);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:305]: >ERR_AmbigQM [](start = 79, length = 11)

Same here, or just explaining what `QM` means.

File: src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs [hunk 8]
```diff
@@ -965,6 +965,9 @@ private void CheckRestrictedTypeReceiver(BoundExpression expression, CSharpCompi
                             case BoundUnconvertedSwitchExpression { Type: { } naturalType } switchExpr:
                                 _ = ConvertSwitchExpression(switchExpr, naturalType ?? CreateErrorType(), conversionIfTargetTyped: null, diagnostics);
                                 break;
+                            case BoundUnconvertedConditionalOperator { Type: { } naturalType } conditionalExpr:
+                                _ = ConvertConditionalExpression(conditionalExpr, naturalType ?? CreateErrorType(), conversionIfTargetTyped: null, diagnostics);
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Invocation.cs:969]: >naturalType [](start = 82, length = 11)

`naturalType` cannot be null here. Did you mean to use `var` above?

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs [hunk 9]
```diff
@@ -3731,119 +3730,65 @@ private BoundExpression BindConditionalOperator(ConditionalExpressionSyntax node
...
+            BoundExpression falseExpr = BindValue(whenFalse, diagnostics, BindValueKind.RValue);
+            HashSet<DiagnosticInfo> useSiteDiagnostics = null;
+            TypeSymbol type = BestTypeInferrer.InferBestTypeForConditionalOperator(trueExpr, falseExpr, this.Conversions, out bool hadMultipleCandidates, ref useSiteDiagnostics);
+            diagnostics.Add(node, useSiteDiagnostics);
+            bool targetTyped;
+            if (type is null)
             {
-                valKind |= BindValueKind.RefersToLocation;
+                noCommonTypeError = hadMultipleCandidates ? ErrorCode.ERR_AmbigQM : ErrorCode.ERR_InvalidQM;
+                targetTyped = Compilation.LanguageVersion >= MessageID.IDS_FeatureTargetTypedConditional.RequiredVersion();
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:3750]: This is somewhat unfortunate, as we're not going to provide a langversion diagnostic. We should consider always target-typing, and just issuing an error if the language version isn't correct.

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 10]
```diff
@@ -124,8 +125,14 @@ public Conversion ClassifyImplicitConversionFromExpression(BoundExpression sourc
             // The switch expression conversion is "lowest priority", so that if there is a conversion from the expression's
             // type it will be preferred over the switch expression conversion.  Technically, we would want the language
             // specification to say that the switch expression conversion only "exists" if there is no implicit conversion
-            // from the type, and we accomplish that by making it lowest priority.
-            return GetSwitchExpressionConversion(sourceExpression, destination, ref useSiteDiagnostics);
+            // from the type, and we accomplish that by making it lowest priority.  The same for the conditional expression
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs:128]: >same for [](start = 88, length = 8)

`is true for` or `goes for`

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs [hunk 11]
```diff
@@ -964,6 +971,22 @@ private Conversion GetSwitchExpressionConversion(BoundExpression source, TypeSym
             }
         }
 
+        private Conversion GetConditionalExpressionConversion(BoundExpression source, TypeSymbol destination, ref HashSet<DiagnosticInfo> useSiteDiagnostics)
+        {
+            if (!(source is BoundUnconvertedConditionalOperator conditionalOperator))
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs:976]: An excellent of example of why `not` should be the same and `!` :)

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs [hunk 12]
```diff
@@ -2101,6 +2101,15 @@ private static bool IsUserDefinedTrueOrFalse(BoundUnaryOperator @operator)
                             break;
                     }
                 }
+                else if (highestBoundExpr is BoundConditionalOperator { WasTargetTyped: true } cond)
+                {
+                    type = cond.NaturalTypeOpt;
+                    convertedType = cond.Type;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:2107]: Nit: both of these could have been pulled out via declaration patterns.

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs [hunk 13]
```diff
@@ -1667,10 +1667,18 @@ private void ReportIfUnused(LocalFunctionSymbol symbol)
 
         public override BoundNode VisitLocal(BoundLocal node)
         {
+            LocalSymbol localSymbol = node.LocalSymbol;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:1670]: Is this related to target-typed conditional?

File: src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs [hunk 14]
```diff
@@ -2564,31 +2564,44 @@ public override BoundNode VisitContinueStatement(BoundContinueStatement node)
...
+        }
+
         public override BoundNode VisitConditionalOperator(BoundConditionalOperator node)
         {
-            var isByRef = node.IsRef;
+            return VisitConditionalOperatorCore(node, node.IsRef, node.Condition, node.Consequence, node.Alternative);
+        }
 
-            VisitCondition(node.Condition);
+        private BoundNode VisitConditionalOperatorCore(
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/FlowAnalysis/AbstractFlowPass.cs:2577]: Consider making this `protected virtual` so that `NullableWalker` doesn't have to override the two separate visit methods, and only has to override this one.

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs [hunk 15]
```diff
@@ -284,7 +284,10 @@ internal IOperation CreateInternal(BoundNode boundNode)
                     return CreateRangeExpressionOperation((BoundRangeExpression)boundNode);
                 case BoundKind.SwitchSection:
                     return CreateBoundSwitchSectionOperation((BoundSwitchSection)boundNode);
+                case BoundKind.UnconvertedConditionalOperator:
+                    throw ExceptionUtilities.Unreachable;
                 case BoundKind.UnconvertedSwitchExpression:
+                    throw ExceptionUtilities.Unreachable;
```
Comment (333fred) [inline src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:290]: Not sure how this would be unreachable. What if someone speculates on a switch expression by itself, and then attempts to get the IOperation tree for it? Please add some tests. Same for unconverted conditional operators.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IDeclarationExpression.cs [hunk 16]
```diff
@@ -106,7 +107,7 @@ public void M2(out int i)
     Predecessors: [B4]
     Statements (0)
 ";
-            VerifyFlowGraphAndDiagnosticsForTest<BlockSyntax>(source, expectedFlowGraph, expectedDiagnostics);
+            VerifyFlowGraphAndDiagnosticsForTest<BlockSyntax>(source, expectedFlowGraph, expectedDiagnostics, parseOptions: TestOptions.Regular8);
```
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IDeclarationExpression.cs:110]: If this affects the output, please add verification for what it changes as well.

File: src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringExpression.cs [hunk 17]
```diff
@@ -1414,7 +1415,7 @@ public void M(bool a, string b, string c, string c2, string p)
                 Diagnostic(ErrorCode.ERR_NoImplicitConv, "a ? b : c").WithArguments("string", "int").WithLocation(8, 20)
             };
 
-            VerifyFlowGraphAndDiagnosticsForTest<BlockSyntax>(source, expectedFlowGraph, expectedDiagnostics);
+            VerifyFlowGraphAndDiagnosticsForTest<BlockSyntax>(source, expectedFlowGraph, expectedDiagnostics, parseOptions: TestOptions.Regular8);
```
Comment (333fred) [inline src/Compilers/CSharp/Test/IOperation/IOperation/IOperationTests_IInterpolatedStringExpression.cs:1418]: Same here.

File: src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs [hunk 18]
```diff
@@ -70147,14 +70141,14 @@ public void StructField_Default_NoType()
...
     static void F()
     {
-        _ = default/*T:<null>!*/.ToString();
+        _ = default/*T:!*/.ToString();
     }
 }";
             var comp = CreateCompilation(source, options: WithNonNullTypesTrue());
             comp.VerifyDiagnostics(
                 // (5,33): error CS0023: Operator '.' cannot be applied to operand of type 'default'
                 //         _ = default/*T:<null>!*/.ToString();
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/NullableReferenceTypesTests.cs:70150]: Comment needs to be updated.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs [hunk 19]
```diff
@@ -0,0 +1,246 @@
...
+
+            // User - defined implicit conversions
+            TestConditional("b ? GetB() : GetC()", "X", null);
+
+            // Anonymous function conversions
+            TestConditional("b ? a=>a : b=>b", "Del", null);
+
+            // Method group conversions
+            TestConditional("b ? M1 : M2", "Del", null);
+        }
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs:69]: Consider adding success case tests for implicit dynamic conversion (operands are completely unrelated)

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs [hunk 20]
```diff
@@ -0,0 +1,246 @@
...
+{
+    /// <summary>
+    /// Test binding of the target-typed conditional (aka ternary) operator.
+    /// </summary>
+    public class TargetTypedConditionalOperatorTests : CSharpTestBase
+    {
+        [Fact]
+        public void TestImplicitConversions_Good()
+        {
+            // Implicit constant expression conversions
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs:27]: Consider testing something that would succeed now, like: `double d = b ? -1L : 1UL;`

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs [hunk 21]
```diff
@@ -0,0 +1,246 @@
...
+            TestConditional("b ? 1 : 2", "System.Int16", "System.Int32",
+                // (6,26): error CS0266: Cannot implicitly convert type 'int' to 'short'. An explicit conversion exists (are you missing a cast?)
+                //         System.Int16 t = b ? 1 : 2;
+                Diagnostic(ErrorCode.ERR_NoImplicitConvCast, "b ? 1 : 2").WithArguments("int", "short").WithLocation(6, 26)
+                );
+
+            // Implicit reference conversions
+            TestConditional("b ? GetB() : GetC()", "A", null);
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs:35]: Consider testing more varieties of reference conversions. For example, nested variant conversions, type parameters, array element type conversions, etc.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs [hunk 22]
```diff
@@ -0,0 +1,246 @@
...
+using Microsoft.CodeAnalysis.Test.Utilities;
+using Roslyn.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests
+{
+    /// <summary>
+    /// Test binding of the target-typed conditional (aka ternary) operator.
+    /// </summary>
+    public class TargetTypedConditionalOperatorTests : CSharpTestBase
```
Comment (333fred) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs:22]: Consider adding tests for pointer conversions.

**Comment by 333fred** [comment]
Done review pass (commit 10)

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs (hunk 4 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs:25]: Those are tests that will become "good" tests if we permit target typing even when there is a natural type.  I think that change is likely after we discuss it at an upcoming LDM.  This is the compatible version of the spec until we have that discussion.

---
In reply to: [410355611](https://github.com/dotnet/roslyn/pull/40604#discussion_r410355611) [](ancestors = 410355611)

File: src/Compilers/CSharp/Portable/Errors/MessageID.cs [hunk 23]
```diff
@@ -301,6 +302,7 @@ internal static LanguageVersion RequiredVersion(this MessageID feature)
                 case MessageID.IDS_FeatureLocalFunctionAttributes: // syntax check
                 case MessageID.IDS_FeatureExternLocalFunctions: // syntax check
                 case MessageID.IDS_FeatureMemberNotNull:
+                case MessageID.IDS_FeatureTargetTypedConditional:  // changes semantics
```
[the remark this answers is not in the document]
Comment (gafter) [reply src/Compilers/CSharp/Portable/Errors/MessageID.cs:305]: The relaxation is in semantic analysis, not in the syntax. In any case, it is likely there will be a change in semantics for existing program; see [https://github.com/dotnet/csharplang/pull/3363](https://github.com/dotnet/csharplang/pull/3363) under **Open Issues**.

---
In reply to: [409942196](https://github.com/dotnet/roslyn/pull/40604#discussion_r409942196) [](ancestors = 409942196)

File: src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs (hunk 5 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:261]: No.  In debug builds we do that for locals and parameters.

---
In reply to: [410393631](https://github.com/dotnet/roslyn/pull/40604#discussion_r410393631) [](ancestors = 410393631)

File: src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs (hunk 9 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/Binder_Operators.cs:3750]: The diagnostics for the error cases do not work out nicely that way.

---
In reply to: [410417426](https://github.com/dotnet/roslyn/pull/40604#discussion_r410417426) [](ancestors = 410417426)

File: src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs (hunk 11 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Binder/Semantics/Conversions/ConversionsBase.cs:976]: It cannot be, except perhaps as a special case at the top level of the pattern immediately beneath an *is-pattern* operator. Even that would undermine the applicability of the demorgan's law transformation for patterns. Cyrus has queued that up for LDM discussion.

---
In reply to: [410419596](https://github.com/dotnet/roslyn/pull/40604#discussion_r410419596) [](ancestors = 410419596)

File: src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs (hunk 12 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Compilation/CSharpSemanticModel.cs:2107]: Yes, at the expense of two more declared variables that we do not need.

---
In reply to: [410425048](https://github.com/dotnet/roslyn/pull/40604#discussion_r410425048) [](ancestors = 410425048)

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs (hunk 13 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:1670]: Only tangentially, as I noticed a cascaded diagnostic in added tests.

---
In reply to: [410433785](https://github.com/dotnet/roslyn/pull/40604#discussion_r410433785) [](ancestors = 410433785)

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 15 above)
Comment (gafter) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:290]: There is no API that I can find to speculate on just an expression. Please be more specific about the scenario you'd like tested.

---
In reply to: [410435964](https://github.com/dotnet/roslyn/pull/40604#discussion_r410435964) [](ancestors = 410435964)

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs (hunk 19 above)
Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs:69]: The *implicit dynamic conversion* is from an expression of type `dynamic` to any type.  If one of the arguments were of type `dynamic`, then it would not be target-typed.  If the target were of type `dynamic`, then that is the same as the target type being of type `object`.

---
In reply to: [410449513](https://github.com/dotnet/roslyn/pull/40604#discussion_r410449513) [](ancestors = 410449513)

**Comment by gafter** [comment]
@333fred I think I've responded to all of your comments. Do you have any others?

@dotnet/roslyn-compiler @agocke May I please have a review from a Sr engineer?

File: src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs (hunk 15 above)
Comment (333fred) [reply src/Compilers/CSharp/Portable/Operations/CSharpOperationFactory.cs:290]: Nothing would stop someone from creating an `ExpressionStatement` with just a ternary in it and speculating on that statement.

---
In reply to: [410480904](https://github.com/dotnet/roslyn/pull/40604#discussion_r410480904) [](ancestors = 410480904,410435964)

**Comment by gafter** [comment]
> I did comment on the test for speculation, but I don't think that's blocking as it's a somewhat niche corner.

Please be much more specific about the scenario you'd like tested.  The code you commented on isn't related to the public APIs that speculate.

File: src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs [hunk 24]
```diff
@@ -1667,10 +1667,18 @@ private void ReportIfUnused(LocalFunctionSymbol symbol)
...
+            if ((localSymbol as SourceLocalSymbol)?.IsVar == true && localSymbol.ForbiddenZone?.Contains(node.Syntax) == true)
```
Comment (agocke) [inline src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:1671]: I must have missed a PR... what's `ForbiddenZone`?

Comment (gafter) [reply src/Compilers/CSharp/Portable/FlowAnalysis/DefiniteAssignment.cs:1671]: It is a region of syntax where you're not allowed to reference the variable.  For example inside the initialization of a `var`-declared local, or inside the invocation that declares an `out var` argument.

File: src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs [hunk 25]
```diff
@@ -0,0 +1,272 @@
...
+    {
+        [Fact]
+        public void TestImplicitConversions_Good()
+        {
+            // NOTE: Some of these are currently error cases, but they would become accepted (non-error) cases
+            // if we extend the spec to permit target typing even when there is a natural type.  Until then,
+            // they are error cases but included here for convenience.
+
+            // Implicit constant expression conversions
+            TestConditional("b ? 1 : 2", "System.Int16", "System.Int32",
```
Comment (agocke) [inline src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs:32]: Could you add a few regression tests for cases where there is no target-type? Cases that may be interesting:

```C#
var x = b ? 1 : 2; // x : int
var x = b ? GetB() : GetC() // error, no common type
var x = b ? $""x"" : $""x"" // x : string
```

Comment (gafter) [reply src/Compilers/CSharp/Test/Semantic/Semantics/TargetTypedConditionalOperatorTests.cs:32]: This file adds target-typing tests. Existing scenarios with a var-declared variable are tested in `ConditionalOperatorTests.cs`. See line 28 there for the first case.
The second scenario is tested at line 95.
Line 119 demonstrates the third case.


---
In reply to: [412552731](https://github.com/dotnet/roslyn/pull/40604#discussion_r412552731) [](ancestors = 412552731)
